draft-ietf-tls-tls13.md

```diff
@@ -3588,6 +3548,127 @@ appropriate application traffic key as described in {{updating-traffic-keys}}.
 
+## 0-RTT and Anti-Replay {#replay-time}
+
+As noted in {{zero-rtt-data}}, unlike 1-RTT data, TLS does not provide
+inherent replay protections for 0-RTT data. Instead, it provides
+mechanisms which allow a server to implement a number of limited
+server-side anti-replay defenses. Servers SHOULD implement
+either Single-Use Tickets {{single-use-tickets}} or
+Client Hello Recording {{client-hello-recording}}
+as described
+below, and if not, SHOULD implement the stateless mechanism
+described in {{stateless-anti-replay}}.
+See {{replay-0rtt}} for more information on the limitations
+of these mechanisms.
+
+
+### Single-Use Tickets
+
+The simplest form of anti-replay defense is for the server to only
+allow each session ticket once. In order to implement this, the server
+maintains a database of all outstanding valid tickets; deleting each
+ticket from the database as it is used. If an unknown ticket is
+provided, the server falls back to a full handshake as normal.
+
+If the tickets are not self-contained but rather are database keys,
+and these PSKs are deleted upon use, then connections established
+using one PSK enjoy forward security with respect to other PSKs
+established on the same connection. This is a security advantage for
+all 0-RTT data and for PSK usage when PSK is used without DH.
+
+Because this mechanism requires sharing the session database between
+server nodes, it may be hard to achieve high rates of PSK and and
+0-RTT success when compared with self-encrypted tickets which do not
+require consistent server-side storage for basic functionality but
+only for 0-RTT anti-replay.
+
+
+### Client Hello Recording
+
+An alternative form of anti-replay is to record each ClientHello or a
+unique value derived from the ClientHello and reject
+duplicates. However, recording all ClientHellos causes state to grow
+without bound, so in practice the server must instead record
+ClientHellos within a given time window based on the
+"obfuscated_ticket_age" value provided by the client.
+
+In order to implement this mechanism, a server needs to store the
+following values, either locally or by encoding them in the ticket:
+
+- The time that the server generated the session ticket.
+- The estimated round trip time between the client and server;
+  this can be estimated by measuring the time between sending
+  the Finished message and receiving the first message in the
+  client's second flight, or potentially using information
+  from the operating system.
+- The "ticket_age_add" parameter from the NewSessionTicket message in
+  which the ticket was established.
+
+The server can determine the client's view of the age of the ticket by
+subtracting the ticket's "ticket_age_add value" from the
+"obfuscated_ticket_age" parameter in the client's "pre_shared_key"
+extension. The server can determine the approximate time that the
+client sent the ClientHello as:
+
+~~~~
+    creation time + (client's view - RTT estimate/2)
```

Comment:
We seem to be having trouble reasoning about how this mechanism works, which is probably my fault for bad writing. In any case, the idea here is supposed to be that when you receive a CH, you need to distinguish three cases:

1. I know it's not a replay, because if I saw it I would still have remembered it.
2. I know it's a replay, in which case I should hard fail.
3. I'm not sure if it's a replay, in which case I force you back to 1-RTT.

Let's pretend that the client didn't have an age value in the CH and I have a window of 10 seconds. So, if I receive a CH at T=0, then an attacker can replay it by waiting 11 seconds, by which time I will have flushed it and I will accept the replay. So, the time of receipt is not itself enough.

Conversely say that the client stuffs its view of age in absolute time in the CH and I have a window that is "I keep any packet which claims to be within 10 seconds of the current time either way". If I receive a CH[T=0] at T=0, then I accept it because my window is [-10,10]. Now, if that packet is replayed at T=11, then my window is now [1, 21] and so I have no information about whether this is a replay or not.

Unless I'm badly confused, if the client's clock is crazily wrong here, then what happens is that they get forced into 1-RTT, but it won't cause me to accept a CH as if it had not been replayed when it had, because I'm always retaining any CH with a claimed time within my current window.

The situation is more subtle when the client is reporting relative time, so before discussing that, perhaps we should make sure we all agree that the logic above is right.